File: algorithms/dynamic-programming/edit-distance/c/editdistance.c
#include <stdio.h>
#include <string.h>

int min(int a, int b, int c) {
    int m = a;
    if (b < m) m = b;
    if (c < m) m = c;
    return m;
}

int edit_distance(const char *s1, const char *s2) {
    int m = strlen(s1);
    int n = strlen(s2);

    int dp[m + 1][n + 1];

    for (int i = 0; i <= m; i++)
        dp[i][0] = i;
    for (int j = 0; j <= n; j++)
        dp[0][j] = j;

    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            int cost = (s1[i - 1] != s2[j - 1]) ? 1 : 0;
            dp[i][j] = min(
                dp[i - 1][j] + 1,
                dp[i][j - 1] + 1,
                dp[i - 1][j - 1] + cost
            );
        }
    }

    return dp[m][n];
}

int main() {
    printf("%d\n", edit_distance("kitten", "sitting")); // 3
    return 0;
}

File: algorithms/dynamic-programming/edit-distance/rust/edit_distance.rs
use std::cmp;

pub fn edit_distance(s1: &str, s2: &str) -> usize {
    let m = s1.len();
    let n = s2.len();
    let s1_bytes = s1.as_bytes();
    let s2_bytes = s2.as_bytes();

    let mut dp = vec![vec![0usize; n + 1]; m + 1];

    for i in 0..=m {
        dp[i][0] = i;
    }
    for j in 0..=n {
        dp[0][j] = j;
    }

    for i in 1..=m {
        for j in 1..=n {
            let cost = if s1_bytes[i - 1] != s2_bytes[j - 1] { 1 } else { 0 };
            dp[i][j] = cmp::min(
                cmp::min(dp[i - 1][j] + 1, dp[i][j - 1] + 1),
                dp[i - 1][j - 1] + cost,
            );
        }
    }

    dp[m][n]
}

fn main() {
    println!("{}", edit_distance("kitten", "sitting")); // 3
}
